src/dune_engine/build_system.ml

```diff
@@ -1140,29 +1151,28 @@ let get_rule_or_source t path =
       let loc = Rule_fn.loc () in
       no_rule_found t ~loc path
   else if Path.exists path then
-    Memo.Build.return (Source (Cached_digest.file path))
+    let+ d = Cached_digest.source_or_external_file path in
+    Source d
   else
     let loc = Rule_fn.loc () in
     User_error.raise ?loc
       [ Pp.textf "File unavailable: %s" (Path.to_string_maybe_quoted path) ]
 
 let all_targets t =
-  let root = File_tree.root () in
-  Context_name.Map.fold t.contexts
-    ~init:(Memo.Build.return Path.Build.Set.empty) ~f:(fun ctx acc ->
-      File_tree.Dir.fold root ~traverse:Sub_dirs.Status.Set.all ~init:acc
-        ~f:(fun dir acc ->
-          let* acc = acc in
+  let* root = File_tree.root () in
+  let module M = File_tree.Dir.Make_map_reduce (Memo.Build) (Path.Build.Set) in
```

Comment:
Since there are multiple `Set` monoids (e.g., based on union and on intersection), it may be clearer to write something like this instead:

```
Make_map_reduce (Memo.Build) (Monoid.Union (Path.Build.Set))
```

Then, I think, you no longer need to bake in the union monoid instance into `Set` as the "blessed" monoid.

Reply:
Btw, there is a similar concern about manual vs automated floating of the module `M`.